io/lab: Explain the need for loops when using read/write (#148)
open-education-hub/operating-systems: content/chapters/io/lab/content/file-descriptors.md

Notice that the code doesn't simply call `read(fd, buff, bytes_to_read)`.
Instead, it uses a `while` loop to read data from the file.

> The [return value of `read()`](https://man7.org/linux/man-pages/man2/read.2.html#RETURN_VALUE) may differ from `bytes_to_read` if there are not enough bytes available or if the operation is interrupted by a signal.

Reviewer: ```suggestion
The [return value of `read()`](https://man7.org/linux/man-pages/man2/read.2.html#RETURN_VALUE) may be less than `bytes_to_read` if there are not enough bytes available or if the operation is interrupted by a signal.
```

> The return value of `read()` may differ from _bytes_to_read_ if there are not enough bytes available or if the operation is interrupted by a signal.
> A return value between **0** and _bytes_to_read_ is not enough to decide whether we should stop reading.

Reviewer: ```suggestion
The return value of `read()` may differ from `bytes_to_read` if there are not enough bytes available or if the operation is interrupted by a signal.
A return value between **0** and `bytes_to_read` is not enough to decide whether we should stop reading.
```

To determine this, we make another `read()` call, which will return **0** if the cursor is already at **EOF** (end of file).

> 

Reviewer: I would also add a link to [the "RETURN VALUE" section from `read()`'s man page](https://man7.org/linux/man-pages/man2/read.2.html#RETURN_VALUE)

But don't bother memorising it.
You can find it any time in by typing [`man fopen`](https://man7.org/linux/man-pages/man3/fopen.3.html) in your terminal.

| `fopen()` mode |          `open()` flag          |
| :------------: | :-----------------------------: |
|     `"r"`      |           `O_RDONLY`            |
|     `"w"`      | `O_WRONLY │ O_CREAT │ O_TRUNC`  |
|     `"a"`      | `O_WRONLY │ O_CREAT │ O_APPEND` |
|     `"r+"`     |            `O_RDWR`             |
|     `"w+"`     |  `O_RDWR │ O_CREAT │ O_TRUNC`   |

> |     `"a+"`     |  `O_RDWR │ O_CREAT │ O_APPEND`  |

Reviewer: attaboy
